### DISCUSSION TITLE
OnLaunch always starts a new instance of Uno ?

### DISCUSSION THREAD

**Discussion by BertrandDeSaintBezier** [unoplatform/uno#12933] [Q&A] [answered]
On Windows, I'm running into an issue where it seems like the OnLaunch() function always starts up a new instance of the application, with all variables reverting to their default values.

In the following example, when the application is activated by a uri scheme redirection (defined in the appxmanifest's "Declarations" section), the OnLaunch function is raised. As pointed out in other threads, the [WinUI OnActivated fuction doesn't exist](https://github.com/unoplatform/uno/discussions/10803#discussion-4685853), so I'm trying to workaround this by checking **_why_** the `OnLaunch` function was called. 
```csharp
protected async override void OnLaunched(LaunchActivatedEventArgs args)
{
#if !HAS_UNO
    var activatedArgs = AppInstance.GetCurrent().GetActivatedEventArgs();
    var activationKind = activatedArgs.Kind;
    if (activationKind == ExtendedActivationKind.Protocol) // true when redirecting. Else = "Launch" when starting app
    {
        OnApplicationActivated();
        return;
    }
#endif

    SetupHosting();
    InitializeAndActivateWindow();
}
```
When the application is first started up, the `activationKind` is of type `ExtendedActivationKind.Launch`, and during uri redirection, it correctly identifies the `activationKind` as being of type `ExtendedActivationKind.Protocol`. However, all variables in the `App.cs` class are reverted to default values, such as is the `_window` object being equal to `null`. 

For instance, the `OnApplicationActivated()` function simply invokes an event, passing the received uri as a parameter. The expected behaviour is that this event shouldn't be null, as some other objects have subscribed to it at some point in the application's lifetime. However, when inspecting the value in debug mode, the event is empty, as if the class had been reinstantiated.
```csharp
private void OnApplicationActivated()
{
#if !HAS_UNO
    var activatedEventArgs = AppInstance.GetCurrent().GetActivatedEventArgs();
    if (activatedEventArgs.Data is ProtocolActivatedEventArgs protocolArgs)
    {
        UriSchemeActivated?.Invoke(this, protocolArgs.Uri);
    }
#endif
}
```
I don't know if this is expected behaviour, or if I'm doing something wrong, but a new process seems to be started up, despite the fact that the application identifies this as being a `Protocol` activation and not a `Launch` one. This can further be confirmed by opening the Task Manager, and seeing that a new process has been started for the application. 

To sum things up, my question is : is this expected behaviour ? And more importantly, is there a way that I can receive an activation of  the application without the OnLaunched function creating a whole new instance ?

Thanks !

**Comment by jeromelaban** [comment]
Thanks for the feedback. To be sure of the scope of the question, when you're saying "On Windows", do you mean the `.Windows` head of your solution?

**Reply by BertrandDeSaintBezier** [reply]
Well, I'm not just building for Windows, but yes, I am having this issue on the Windows head right now.

**Reply by BertrandDeSaintBezier** [reply, marked as answer]
Okay so after some digging, this is apparently normal in WinUI3. There is a workaround solution provided in [this repo](https://github.com/andrewleader/RedirectActivationWinUI3Sample/blob/main/RedirectActivationWinUI3Sample/App.xaml.cs) in which the application registers itself with a key and any subsequent launches are checked against it. If it detects that an application instance is already running, the arguments are forwarded to that instance and the new process is killed.

Here's the gist of it :

```csharp
protected async override void OnLaunched(LaunchActivatedEventArgs args)
{
    var activatedArgs = AppInstance.GetCurrent().GetActivatedEventArgs();
    var mainInstance = AppInstance.FindOrRegisterForKey("main");

    if (mainInstance.IsCurrent is not true)
    {
        await mainInstance.RedirectActivationToAsync(activatedArgs);
        Process.GetCurrentProcess().Kill();
        return;
    }
    else
    {
        AppInstance.GetCurrent().Activated += OnApplicationActivated;
    }

    SetupHosting();
    InitializeAndActivateWindow();
}

private void OnApplicationActivated(object? sender, AppActivationArguments args)
{
    // perform activation logic
}
```
